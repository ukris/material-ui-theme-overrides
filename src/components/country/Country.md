Country with defaults
```jsx
    <Country setShowEditor={(content) => console.log(`setShowEditor: ${content}`)} updateCell={(newCellValue) => alert(`Sent to Parent: ${newCellValue}`)} />
```
Country with invalid country defaults to default country
```jsx
    <Country country="zz" setShowEditor={(content) => console.log(`setShowEditor: ${content}`)} updateCell={(newCellValue) => alert(`Sent to Parent: ${newCellValue}`)}/>
```

Country with country with size
```jsx
    <Country country="IN" size="lg" setShowEditor={(content) => console.log(`setShowEditor: ${content}`)} updateCell={(newCellValue) => alert(`Sent to Parent: ${newCellValue}`)}/>
```

Editable Country
```jsx
    <Country country="IN" setShowEditor={(content) => console.log(`setShowEditor: ${content}`)} updateCell={(newCellValue) => alert(`Sent to Parent: ${newCellValue}`)}/>
```