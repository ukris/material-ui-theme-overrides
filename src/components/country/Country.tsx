import React from 'react'
import { WithStyles, withStyles, Theme} from '@material-ui/core/styles'
import clsx from 'clsx'
import TextField from '@material-ui/core/TextField'
import Autocomplete from '@material-ui/lab/Autocomplete'
import ClickAwayListener from '@material-ui/core/ClickAwayListener'

import { countryToFlag, countries } from 'helpers/countries'
import GlobalContext from 'context/global-context'
import { Size } from 'types'
import defaults from 'settings'

import './styles.scss'

const defaultCountryFlag =  countryToFlag(defaults.country)

export interface StyleProps  extends WithStyles<typeof styles> {
   
}

export interface RenderProps extends StyleProps {
    /** Two characters Country Code string */
    country?: string
    /** Font size */
    size?: Size
    handleClick?: (e:any) => void
}

export interface EditorProps extends StyleProps {
    /** Signals if Country can be changed or not */

    /** Two characters Country Code string */
    country?: string

    /** Font size */
    size?: Size

    /** Updates row data in parent component with the purpose of persisting when switching table page */
    updateCell: (newCellValue:any) => void
    setShowEditor: (content:any) => void
}

interface CountryType {
    code: string
    label: string
    phone: string
}

const styles = (theme: Theme):any => {
    const { fontSize } = theme
    return {
        isoRoot: {
            marginTop:'-0.125em',
            marginLeft: '1em',
            fontSize: fontSize.md
        },
        isoVal: {
            marginLeft: '0.25rem'
        },
        option: {
            fontSize: 15,
            '& > span': {
              marginRight: 10,
              fontSize: 18,
            },
        },
        input: {
            widht: '100%',
            opacity: 0
        }
    }
}

function CountryCell(props: RenderProps) {
    const { classes, country = defaults.country, size = "md", handleClick} = props
    let countryFlag = countryToFlag(country)
    let countryVal = country
    if (!countryFlag) {
        countryFlag = defaultCountryFlag
        countryVal = defaults.country
    }

    return (
        <span className={clsx(classes.isoRoot)} onClick={(e:any) => handleClick && handleClick(e)} data-test="table-country">  
            <span data-test="table-country-flag">{countryFlag}</span>
            <span
                className={clsx(classes.isoVal)}
                data-test="table-country-code"
            >
                {countryVal}
            </span>
        </span>
    )
}

function  Country(props: EditorProps) {
    const { classes, country = defaults.country, size = "md", updateCell, setShowEditor} = props
    const [countryCode, setCountryCode] = React.useState(country)

    let countryFlag = countryToFlag(countryCode)
    let countryVal = countryCode
    if (!countryFlag) {
        countryFlag = defaultCountryFlag
        countryVal = defaults.country
    }

    return (
        <ClickAwayListener onClickAway={()=> {
            setShowEditor(null)
        }}>
        <Autocomplete
            id="country-select-demo"
            style={{ width: 300 }}
            open={true}
            options={countries as CountryType[]}
            classes={{
                option: classes.option,
                input: classes.input
            }}
            autoHighlight
            getOptionLabel={(option) => option.code}
            renderOption={(option) => (
                <React.Fragment>
                    <span data-test="table-country-selectableflags">{countryToFlag(option.code)}</span>
                    {option.label} ({option.code}) +{option.phone}
                </React.Fragment>
            )}
            renderInput={(params) => {
                // @ts-ignore
                const selectedVal: string = params.inputProps.value
                if (selectedVal && selectedVal !== countryCode) {
                    setCountryCode(selectedVal)
                    updateCell && updateCell(selectedVal)
                }

                return (
                    <TextField
                    {...params}
                    label="Choose a country"
                    variant="outlined"
                    inputProps={{
                        ...params.inputProps,
                        autoComplete: 'new-password', // disable autocomplete and autofill
                    }}
                    />
                )
            }}
        />
        </ClickAwayListener>
    )
}

export default withStyles(styles)(Country)

export const RenderCountry = withStyles(styles)(CountryCell)