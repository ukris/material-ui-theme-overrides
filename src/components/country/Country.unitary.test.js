import React from 'react'
import renderer from 'react-test-renderer'
import { mount, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { RenderCountry } from './Country'

configure({ adapter: new Adapter() })


describe('Country | Non-editable | Valid Country Code', () => {
    const component = <RenderCountry country="FR" handleClick={() => false}/>
    const wrapper = mount(component)

    it('renders component correctly', () => {
        const divContainer = wrapper.find('[data-test="table-country"]')
        const countryFlag = wrapper.find('[data-test="table-country-flag"]')
        const countryCode = wrapper.find('[data-test="table-country-code"]')

        expect(divContainer.length).toEqual(1)
        expect(countryFlag.length).toEqual(1)
        expect(countryFlag.props().children).toEqual("ðŸ‡«ðŸ‡·")
        expect(countryCode.length).toEqual(1)
        expect(countryCode.props().children).toEqual("FR")
    });

    it('matches snapshot', () => {
        const tree = renderer.create(component,).toJSON()
        const countryFlag = tree.children[0]
        const countryCode = tree.children[1]

        expect(tree).toMatchSnapshot();
        expect(countryFlag.type).toEqual('span')
        expect(countryFlag.children[0]).toEqual("ðŸ‡«ðŸ‡·")
        expect(countryCode.type).toEqual('span')
        expect(countryCode.children[0]).toEqual('FR')
    });
})

describe('Country | Non-editable | Invalid Country Code', () => {
    const component = <RenderCountry country="zz" handleClick={() => false}/>
    const wrapper = mount(component)

    it('renders component with default flag', () => {
        const divContainer = wrapper.find('[data-test="table-country"]')
        const countryFlag = wrapper.find('[data-test="table-country-flag"]')
        const countryCode = wrapper.find('[data-test="table-country-code"]')

        expect(divContainer.length).toEqual(1)
        expect(countryFlag.length).toEqual(1)
        expect(countryFlag.props().children).toEqual("ðŸ‡ºðŸ‡¸")
        expect(countryCode.length).toEqual(1)
        expect(countryCode.props().children).toEqual("US")
    });

    it('matches snapshot', () => {
        const tree = renderer.create(component,).toJSON()
        const countryFlag = tree.children[0]
        const countryCode = tree.children[1]

        expect(tree).toMatchSnapshot();
        expect(countryFlag.type).toEqual('span')
        expect(countryFlag.children[0]).toEqual("ðŸ‡ºðŸ‡¸")
        expect(countryCode.type).toEqual('span')
        expect(countryCode.children[0]).toEqual('US')
    });
})

describe('Country | Non-editable | Default Country Code', () => {
    const component = <RenderCountry handleClick={() => false}/>
    const wrapper = mount(component)

    it('renders component correctly', () => {
        const divContainer = wrapper.find('[data-test="table-country"]')
        const countryFlag = wrapper.find('[data-test="table-country-flag"]')
        const countryCode = wrapper.find('[data-test="table-country-code"]')

        expect(divContainer.length).toEqual(1)
        expect(countryFlag.length).toEqual(1)
        expect(countryFlag.props().children).toEqual("ðŸ‡ºðŸ‡¸")
        expect(countryCode.length).toEqual(1)
        expect(countryCode.props().children).toEqual("US")
    });

    it('matches snapshot', () => {
        const tree = renderer.create(component,).toJSON()
        const countryFlag = tree.children[0]
        const countryCode = tree.children[1]

        expect(tree).toMatchSnapshot();
        expect(countryFlag.type).toEqual('span')
        expect(countryFlag.children[0]).toEqual("ðŸ‡ºðŸ‡¸")
        expect(countryCode.type).toEqual('span')
        expect(countryCode.children[0]).toEqual('US')
    });
})
