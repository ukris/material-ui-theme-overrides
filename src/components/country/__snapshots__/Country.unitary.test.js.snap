// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Country | Non-editable | Default Country Code matches snapshot 1`] = `
<span
  className="CountryCell-isoRoot-1"
  data-test="table-country"
  onClick={[Function]}
  style={
    Object {
      "fontSize": "1rem",
    }
  }
>
  <span
    data-test="table-country-flag"
  >
    ðŸ‡ºðŸ‡¸
  </span>
  <span
    className="CountryCell-isoVal-2"
    data-test="table-country-code"
  >
    US
  </span>
</span>
`;

exports[`Country | Non-editable | Invalid Country Code matches snapshot 1`] = `
<span
  className="CountryCell-isoRoot-1"
  data-test="table-country"
  onClick={[Function]}
  style={
    Object {
      "fontSize": "1rem",
    }
  }
>
  <span
    data-test="table-country-flag"
  >
    ðŸ‡ºðŸ‡¸
  </span>
  <span
    className="CountryCell-isoVal-2"
    data-test="table-country-code"
  >
    US
  </span>
</span>
`;

exports[`Country | Non-editable | Valid Country Code matches snapshot 1`] = `
<span
  className="CountryCell-isoRoot-1"
  data-test="table-country"
  onClick={[Function]}
  style={
    Object {
      "fontSize": "1rem",
    }
  }
>
  <span
    data-test="table-country-flag"
  >
    ðŸ‡«ðŸ‡·
  </span>
  <span
    className="CountryCell-isoVal-2"
    data-test="table-country-code"
  >
    FR
  </span>
</span>
`;
